### PR TITLE
Protocol E016H

### PR DESCRIPTION
**Pull request by goebish**
Add protocol for Eachine E016H
Tested, working.

Protocol name: E016H
Channel mapping:
Channel 5: flip
Channel 6: headless
Channel 7: RTH
Channel 8: emergency stop

Documentation update incoming (I know I'm 2 or 3 protocols late already ...)

### CODE REVIEW

File: src/protocol/e016h_nrf24l01.c [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+#ifdef EMULATOR
+#define USE_FIXED_MFGID
+#define E016H_BIND_COUNT 4
+#define dbgprintf printf
+#else
+#define E016H_BIND_COUNT 750  // 3 seconds
+//printf inside an interrupt handler is really dangerous
+//this shouldn't be enabled even in debug builds without explicitly
+//turning it on
+#define dbgprintf if(0) printf
```
Comment (PhracturedBlue) [inline src/protocol/e016h_nrf24l01.c]: Lint Error:
```
38:  Should have a space between // and comment  
39:  Should have a space between // and comment  
40:  Should have a space between // and comment  
41:  Missing space before ( in if(  
```

File: src/protocol/e016h_nrf24l01.c [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        packet[7] = val >> 8;
+        packet[8] = val & 0xff;
+        // flags
+        packet[1] |= GET_FLAG(CHANNEL8, FLAG_STOP)
+                  |  (can_flip ? GET_FLAG(CHANNEL5, FLAG_FLIP) : 0);
+        packet[3] |= GET_FLAG(CHANNEL6, FLAG_HEADLESS)
+                  |  GET_FLAG(CHANNEL7, FLAG_RTH);
+        packet[7] |= FLAG_HIGHRATE;
+        // frequency hopping
+        NRF24L01_WriteReg(NRF24L01_05_RF_CH, hopping_frequency[hopping_frequency_no++ & 0x03]);
```
Comment (PhracturedBlue) [inline src/protocol/e016h_nrf24l01.c]: Lint Error:
```
154:  Line ends in whitespace.  Consider deleting these extra spaces.  
```

File: src/protocol/e016h_nrf24l01.c [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+                  |  GET_FLAG(CHANNEL7, FLAG_RTH);
+        packet[7] |= FLAG_HIGHRATE;
+        // frequency hopping
+        NRF24L01_WriteReg(NRF24L01_05_RF_CH, hopping_frequency[hopping_frequency_no++ & 0x03]);        
+    }
+    // checksum
+    packet[9] = packet[0];
+    for (u8 i=1; i < E016H_PACKET_SIZE-1; i++)
+        packet[9] += packet[i];
+
```
Comment (PhracturedBlue) [inline src/protocol/e016h_nrf24l01.c]: Lint Error:
```
160:  Line ends in whitespace.  Consider deleting these extra spaces.  
```

File: src/protocol/e016h_nrf24l01.c [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    packet[9] = packet[0];
+    for (u8 i=1; i < E016H_PACKET_SIZE-1; i++)
+        packet[9] += packet[i];
+    
+    // transmit packet
+    XN297_Configure(BV(NRF24L01_00_EN_CRC) | BV(NRF24L01_00_CRCO) | BV(NRF24L01_00_PWR_UP));
+    NRF24L01_WriteReg(NRF24L01_07_STATUS, 0x70);
+    NRF24L01_FlushTx();
+    XN297_WritePayload(packet, E016H_PACKET_SIZE);
+
```
Comment (PhracturedBlue) [inline src/protocol/e016h_nrf24l01.c]: Lint Error:
```
166:  Line ends in whitespace.  Consider deleting these extra spaces.  
```

File: src/protocol/e016h_nrf24l01.c [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    }
+    dbgprintf("\r\n");
+#endif
+    if (Model.fixed_id) {
+       for (u8 i = 0, j = 0; i < sizeof(Model.fixed_id); ++i, j += 8)
+           rand32_r(&lfsr, (Model.fixed_id >> j) & 0xff);
+    }
+    // Pump zero bytes for LFSR to diverge more
+    for (u8 i = 0; i < sizeof(lfsr); ++i) rand32_r(&lfsr, 0);
+
```
Comment (PhracturedBlue) [inline src/protocol/e016h_nrf24l01.c]: Lint Error:
```
211:  Line ends in whitespace.  Consider deleting these extra spaces.  
```

File: src/protocol/e016h_nrf24l01.c [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    // Pump zero bytes for LFSR to diverge more
+    for (u8 i = 0; i < sizeof(lfsr); ++i) rand32_r(&lfsr, 0);
+    
+    // tx id
+    rx_tx_addr[0] = 0xa5;
+    rx_tx_addr[1] = 0x00;
+    rx_tx_addr[2] = lfsr >> 24;
+    rx_tx_addr[3] = lfsr >> 16;
+    rx_tx_addr[4] = lfsr >> 8;
+
```
Comment (PhracturedBlue) [inline src/protocol/e016h_nrf24l01.c]: Lint Error:
```
218:  Line ends in whitespace.  Consider deleting these extra spaces.  
```

File: src/protocol/spi/nrf24l01.c [hunk 7]
```diff
@@ -434,6 +434,8 @@ u8 XN297_WritePayload(u8* msg, int len)
         packet[last++] = crc >> 8;
         packet[last++] = crc & 0xff;
     }
+    if (++last > 32)
```
Comment (goebish) [inline src/protocol/spi/nrf24l01.c]: Weird, I've to add an extra byte to the emulated packet or the last byte of the packet (CRC low nibble) is replaced by CRC high nibble (analyzed the packets with SDR, eg expected CRC is 06EB but 0606 is transmitted). Seems it happens only with some specific packet lengths (10), there's something fishy here.

Comment (goebish) [reply src/protocol/spi/nrf24l01.c]: I think I found the issue, seems like there is a race condition in NRF24L01_WritePayload, CS_HI() can be called before PROTOSPI_xfer has finished its work. Adding an usleep(5) just before CS_HI() fixes the problem. That's not elegant tho, I suppose we can check a flag to make sure the SPI transfer has completed instead ?

File: src/protocol/spi/nrf24l01.c [hunk 8]
```diff
@@ -88,6 +88,7 @@ u8 NRF24L01_WritePayload(u8 *data, u8 length)
     {
         PROTOSPI_xfer(data[i]);
     }
+    usleep(1);  // makes sure SPI transfer has completed
```
Comment (howard0su) [inline src/protocol/spi/nrf24l01.c]: there is no flag to check if the transfer is finished? sleep 1 us is not safe enough.

Comment (goebish) [reply src/protocol/spi/nrf24l01.c]: I don't know if there's a flag, I'm sure all we need is a waiting a couple of cycles, a few nops would probably be enough, I've seen this happen only once.

Comment (goebish) [reply src/protocol/spi/nrf24l01.c]: Well, it's already supposed to wait already 
/* Wait for transfer finished. */
while (!(SPI_SR(spi) & SPI_SR_RXNE));

Comment (goebish) [reply src/protocol/spi/nrf24l01.c]: Well, it's already supposed to wait
https://github.com/DeviationTX/libopencm3/blob/7e1d3daa113e1ae21f308759e50be2f89361d90f/lib/stm32/common/spi_common_all.c#L236

**Pull request merged**